### ISSUE TITLE
Can't create/mount VHDs on ZFS volumes, using Disk Management

### ISSUE THREAD

**Issue by vmlemon** [openzfsonwindows/ZFSin#101]
When trying to use the Disk Management MMC snap-in, to create a disk image (doesn't matter if it's in VHD, or VHDX format, fixed size, or dynamically expanding), on a ZFS volume, the error "The specified path is invalid" is produced, and no disk image is created.

Additionally, if a VHD is copied to the ZFS volume, after being created on another volume, and an attempt is made to mount it, using Windows Explorer, the error "Sorry, there was a problem mounting the file" is produced, and the mount process fails.

The case of the filename doesn't seem to matter, and I'm testing on a volume with the default case sensitivity mode, and deduplication enabled.

Using OpenZFS 0.15, on Windows 10 Pro Build 16299.rs3_release.170928-15434.

**Comment by lundman** [comment]
Interesting - I have never made a VHD, but I will give it a go and see what it complains about.

**Comment by jheuking** [comment]
ok, I just tested it and it works here.
Guess that is the same issue I had with downloading files with Microsoft Edge, which should be fixed now, but is not yet in released binaries.
Just to make sure, we mean the same things, i did:
zvol create -V 1GB tank/vol
convert MBR and format NTFS
Disk Management: Create VHD test.vhd with size 200MB on tank/vol

**Comment by lundman** [comment]
I thought he meant create vhd file on dataset? Or was it ZVOL?

**Comment by jheuking** [comment]
oh damn, you are right, that doesn't work here either

**Comment by vmlemon** [comment]
> oh damn, you are right, that doesn't work here either
> got confused by the ZFS Volume

Yeah - not a block ZVOL, but rather, creating the VHD file, in a random directory, on a standard dataset. (I'm thinking of a "volume", as in a mountpoint).

**Comment by jheuking** [comment]
only suspicious thing I see is ` user_fs_request: unknown class 0x9c040`

**Comment by lundman** [comment]
![screen shot 2019-02-12 at 0951 37 2](https://user-images.githubusercontent.com/637168/52603848-eeab4d80-2eab-11e9-913a-dfb97d3e11e6.png)

Now - I wonder if it is the difference in getting name, normalized or not - I suspect we are supposed to reply with full path in one of the cases.

**Comment by lundman** [comment]
As compared to when created on NTFS;

![screen shot 2019-02-12 at 1019 35 2](https://user-images.githubusercontent.com/637168/52605076-a6425e80-2eb0-11e9-8874-26cffbc44643.png)

**Comment by lundman** [comment]
Ok, fixing full filenames to be returned, gets it a little further, in this case, we fail (the second time) on sharing access. It has share_read set, but we fail it (compared to ntfs). If I comment out

```
			Status = IoCheckShareAccess(granted_access, IrpSp->Parameters.Create.ShareAccess, FileObject, vp ? &vp->share_access : &dvp->share_access, FALSE);
====>			Status = 0;
			if (!NT_SUCCESS(Status)) {
// snip
				dprintf("%s: denied due to IoCheckShareAccess\n", __func__);
				return Status;
```

We can successfully create a VXD, initialize it, format as NTFS then we die:

*** Assertion failed: parent->avl_child[which_child] == NULL
***   Source File: c:\src\zfsin\zfsin\spl\module\spl\spl-avl.c, line 498
Stack:
```
 	ZFSin!zfs_range_add_reader+0x1ae [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_rlock.c @ 306]	C/C++/ASM
 	ZFSin!zfs_range_lock_reader+0x290 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_rlock.c @ 417]	C/C++/ASM
 	ZFSin!zfs_range_lock+0x16d [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_rlock.c @ 453]	C/C++/ASM
 	ZFSin!zil_lwb_commit+0x99f [c:\src\zfsin\zfsin\zfs\module\zfs\zil.c @ 1570]	C/C++/ASM
 	ZFSin!zil_process_commit_list+0x30e [c:\src\zfsin\zfsin\zfs\module\zfs\zil.c @ 2182]	C/C++/ASM
 	ZFSin!zil_commit_writer+0x111 [c:\src\zfsin\zfsin\zfs\module\zfs\zil.c @ 2318]	C/C++/ASM
 	ZFSin!zil_commit_impl+0x45 [c:\src\zfsin\zfsin\zfs\module\zfs\zil.c @ 2818]	C/C++/ASM
 	ZFSin!zil_commit+0x235 [c:\src\zfsin\zfsin\zfs\module\zfs\zil.c @ 2783]	C/C++/ASM
 	ZFSin!zfs_fsync+0x125 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops.c @ 3143]	C/C++/ASM
 	ZFSin!flush_buffers+0x7d [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops_windows.c @ 4737]	C/C++/ASM
 	ZFSin!fsDispatcher+0xc94 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops_windows.c @ 5944]	C/C++/ASM
```

Don't really want to look at that right now though :)

So, will check why the ShareAccess computations go wrong.

**Comment by lundman** [comment]
Ok, we can call `IoUpdateShareAccess()` twice on a file in certain circumstances. 

We still die in spl-avl rangelocking.

**Comment by lundman** [comment]
Ok, HEAD can now create VHDs on ZFS datasets

**Comment by vmlemon** [comment]
Thanks for looking into this.

I still get the "The specified path is invalid" error, when testing with the 0.16 release. Did those changes manage to make it into this one? (I haven't got a working SDK, and Visual Studio instance on this machine, since it only has a 32GB eMMC for internal storage, and VS doesn't seem to like being installed on anything that's not an internal, NTFS volume).

**Comment by lundman** [comment]
Really? I tested it again this morning without issue, both debug and release installers.

**Comment by vmlemon** [comment]
Somehow, I still received the errors, with the "Vegnest" build, on Windows 10 Pro 16299 - but after removing that build, and installing the test build, from Issue #106, I can confirm that I was able to create a 4MB, dynamically-expanding VHD, on a ZFS dataset created under FreeBSD.

If unmount the VHD, and then double-click the icon, in Windows Explorer, I receive:

![image](https://user-images.githubusercontent.com/35642/53925368-a381f600-4077-11e9-96a1-798d28b1d1fb.png)

And, in Disk Manager, I can confirm that the image gets reattached, anyway:

![image](https://user-images.githubusercontent.com/35642/53925390-be546a80-4077-11e9-8c3b-dec56e1d63f4.png)

In Disk Manager, I can create an MBR, and even format the tiny partition, in FAT16 - which appears, in Windows Explorer, without any crashes, or BSODs!

Thanks once again, for looking into this. :)

**Comment by lundman** [comment]
I'll take a look at the "Couldn't mount file" it's probably just missing an API call (even if it does work in the long run)

**Comment by vmlemon** [comment]
Wouldn't be surprised if the recent security/file-locking fixes haven't made the difference, regarding my prior inability to create/mount images - I kept getting bombarded with the "directory loses permissions, and locks itself" glitch, when running applications, or downloading files, at random times; and, Chrome would often BSOD, when trying to download files, for me...

Not sure if it's my configuration being screwed up, or a bug in the build of Windows that I'm running - although, I've already sacrificed the contents of a 2TB HDD, at the instability altar, in the name of science, whilst testing this stuff. 

(Managed to trash VirtualBox, whilst it was still running, on the host, along with a few 200+GB disk images, on an NTFS partition, trying to install FreeBSD, on a different raw USB disk, since the PHYSICALDRIVE paths changed, from when I originally created the "link VHDK", to do some pool testing - not the fault of ZFSin, though) :(

Found some fun things, with PCSX2 plug-ins, and iTunes (won't let me rip CDs, or download store purchases, without throwing Error -50), too - but, I'll probably log those as their own bugs, when I get time.

**Comment by vmlemon** [comment]
I think _that_ "Couldn't mount..." is by design, though - I receive that, immediately after creating the VHD, presumably since there's no volume/partition table, in the image, that's recognised by an IFS driver, since it obviously just doesn't exist.

**Comment by lundman** [comment]
Oh is it the same if you do the same procedure on NTFS?

**Comment by vmlemon** [comment]
Good question.

If I create a new VHD, of the same spec, on an NTFS drive, Windows Explorer doesn't say anything (the image appears in Disk Management as "Disk 5 : Unknown, 4MB, Not initialised", but there's no icon in "This PC", even if I create an MBR, but not a partition). If I unmount the drive, and double-click the Desktop icon, I get the "...contains partitions that aren't recognisable...", and the image appears as "Disk 4 : Basic, 3MB, Online" (since I dismounted the image, from earlier).

After I dismount the VHD on NTFS, and create a new 4MB image, on ZFS, the image gets created, and a new "Disk 4 : Unknown, 4MB, Not initialised" appears, without a partition table, in Disk Management. The "...contains partitions..." Explorer message doesn't appear (and, there's no other, extraneous messages).

If I dismount the ZFS image, and double-click on it, I get the "...contains partitions...", and it appears, as expected, in Disk Management, ready for me to dismount it, once more. 

With both ZFS/NTS: On a VHD that contains an MBR partition table (haven't tested with GPT, since I'm only using tiny images), if a partition without an FS is created, Windows Explorer will say:

![image](https://user-images.githubusercontent.com/35642/53926825-cd89e700-407c-11e9-9522-5325c878af34.png)

Unless something is trying to automount the very-first created image, immediately after creation, I was only able to get the "Couldn't mount...", if I consciously hadn't applied formatting/partitioning to the image.

-

On the flip side, prior to testing the new, unofficial build, I _definitely_ just couldn't create images, at all, on this system, for some reason, though. Wouldn't be surprised, if it didn't miraculously start working, if I reverted to the previous build, and tried to trace the file system calls.

Need to do more testing, with creating VHDs, on a ZFS dataset, that's contained inside another VHD, on an NTFS volume - but, I can safely say, now, that it works fine, on a 2TB USB drive, containing a full-disk ZFS pool, containing a FreeBSD installation.

**Comment by vmlemon** [comment]
(If it makes a difference, deduplication, and compression are both enabled, on that 2TB ZPool, as are directory-based extended attributes)

**Comment by vmlemon** [comment]
So, in a nutshell - "Couldn't mount", is always supposed to happen, if there's no recognised FS, after double-clicking the VHD, contained on a volume mounted using a well-behaved IFS driver (NTFS, and seemingly, now OpenZFSOnWindows-debug-20190307), that does whatever contortions are necessary, in order to make the disk image subsystem happy.

The verdict is still out, on the other "Couldn't mount", that appears, after a new VHD is born - but, however it occurs, it doesn't massively break the core functionality, in a showstopping manner.

Also, I haven't tried brutalising the VHD infrastructure, by doing stuff like creating massive images, using VHDX, creating fixed-size images, or piling in masses of data (or, merely just running Chrome, in Sandboxie), so I can't testify its long-term stability (although, I know that the standard 0.16 build _can_ withstand Ubuntu/FreeBSD/ReactOS, running in VirtualBox, on dynamically-sized VHDs, with 2GB of RAM assigned to them, on a box with only 4GB installed, and deduplication enabled!)

**Comment by lundman** [comment]
Great, thanks for testing. Also, "friends don't let friends dedup." :)

**Comment by vmlemon** [comment]
Did a little extra testing, before I went to work, for a night shift, and not long, after I returned home...

Managed to create a 40GB, dynamically-expanding VHDX image, with GPT/NTFS formatting, installed "Age of Empires : Castle Siege", from the Microsoft Store on it, and left my laptop running, for 4 hours+, and returned home (at about 12:30am), to discover that it hadn't crashed, or BSODed, with the image continuously-mounted. (The image also has the "Better performance" (write caching enabled) removal policy - haven't tested with "Quick removal", but there doesn't seem to have been any data loss, from taking the image offline, and then making a ZFS snapshot).

Even though it wasn't strictly-necessary (since the underlying storage probably looks like mincemeat, with deduplication/compression - but, miraculously, I have "100% space efficiency"), I ran a defragmentation operation, as well as a file system integrity check - both of which completed, without any crashing, or bogus corruption (I've seen cases, on WinBTRFS, where the image is held open, so that NTFS always gets marked as "dirty", whenever running the check, and a ton of additional I/O gets done).

Actually ran the Age of Empires game, to see what'd happen - it launched, without a hitch, other than complaining about being unable to connect to XBox Live, since I uninstalled the prerequisites for doing that, but that was probably to be expected, and isn't a fault of ZFSin.

Allowed Sandboxie to use the NTFS volume, to store its sandbox, and managed to launch Google Chrome, inside it - all of the processes spawn, as expected, despite "`SBIE2101 Object name not found: \BaseNamedObjects\IGFXKMDNotifyBatchBuffersComplete, error OpenEvent (C0000022) access=001F0003 initialized=1`" errors, and it asks me if I want to restore pages from a previous browsing session - attempting this, on a bare ZFS dataset used to violently wreck the FS, so it's amazing that there's no crashing, with so many layers of abstraction. Didn't restore the session, but I managed to load the YouTube homepage, and played an 11:52 video, completely (including letting it automatically play through, to a recommendation, partially), without any crashes. Also downloaded 3 large files (178MB, 491MB, and 389MB), from http://bitsavers.trailing-edge.com/bits/SGI/mips/cd/irix_set1/, to the root of the volume (the last 2 were done, concurrently, to encourage fragmentation/write-interlacing to form, and to see how the ZFS compression/deduplication, and snapshotting would cope). and was amazed to see that all of them had successfully completed.

Set iTunes to use a new directory, in the root of the NTFS drive, downloaded 3 previously-purchased Store tracks - which seems to work fine (this currently fails, with Error -50, on bare ZFS), and ripped a 12 track audio CD into AIFF/WAVE formats (both sets of files are 525MB, regardless of format, leaving a combined 1.02GB of data), without problems; and, also used Windows Media Player, to rerip the same CD, to WAVE format, in a different directory (also 525MB of data, but the deduplication engine treats it as new data, and reduces the duplicates count, since the iTunes, and WMP metadata handling implementations don't work identically).

Let the USBPCapCMD utility recursively write USB traffic, from the hard disk holding the ZFS pool (and NTFS image), into a trace file, on the root of the NTFS volume (which, predictably baloons quickly, even when the disk is supposedly idle, and when running "zfs info", and lets Task Manager produce some funny sawtooth disk traffic graphs, as well as pegging the CPU at 100% (but, the idle time almost perfectly mirrors the disk activity, in shape, occasionally), and the RAM utilisation near 3.4GB/3.8GB - which doesn't even make a dent, in performance, amazingly, despite the low-spec hardware). Snapshotting the ZPool works fine, without skipping a beat (no hanging, or stalling), let alone crashing, after generating nearly 800MB+ of highly-redundant data.

Encrypted the new NTFS image, using BitLocker's "new encryption mode", without issues - deduplication ratio decreased (to 1.19x), as expected, when writing effectively random data, despite the existence of several snapshots, created before, during, and after the encryption process; later, decrypted the image, and observed that the ratio increased (to 1.32x), all without any crashes.

There are probably other tests that I could try (running VMs, nesting multiple VHDs, letting the USB capture run, overnight; or using WinImage, to create a disk image of a 128GB MicroSD card, on a larger VHD?), but I'm confident enough, so far, to say that this bug can now be closed, after doing these tests, and observing no major crashes, or instability.

As for dedup - I'm surprised that it's managed to hold up, to such punishment, without completely falling apart, on such crappy hardware, although the main test was merely to see how well VHD handling would work (pretty well, compared to NTFS, on ZVols, surprisingly, when it comes to BSODs). :)
